src/stream/src/executor/row_merge.rs

```diff
@@ -0,0 +1,153 @@
+    pub lhs_mapping: ColIndexMapping,
+    /// Maps input from the rhs to the output.
+    pub rhs_mapping: ColIndexMapping,
+    /// Output schema
+    pub schema: Schema,
+}
+
+impl RowMergeExecutor {
+    pub fn new(
+        ctx: ActorContextRef,
+        lhs_input: Executor,
+        rhs_input: Executor,
+        lhs_mapping: ColIndexMapping,
+        rhs_mapping: ColIndexMapping,
+        schema: Schema,
+    ) -> Self {
+        Self {
+            ctx,
+            lhs_input,
+            rhs_input,
+            lhs_mapping,
+            rhs_mapping,
+            schema,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    pub async fn execute_inner(self) {
+        let lhs_mapping = self.lhs_mapping;
+        let rhs_mapping = self.rhs_mapping;
+        let data_types = self
+            .schema
+            .fields()
+            .iter()
+            .map(|f| f.data_type())
+            .collect::<Vec<_>>();
+
+        {
+            let mut lhs_buffer: Option<StreamChunk> = None;
+            let mut rhs_buffer: Option<StreamChunk> = None;
+            let aligned_stream = barrier_align(
+                self.lhs_input.execute(),
+                self.rhs_input.execute(),
+                self.ctx.id,
+                self.ctx.fragment_id,
+                self.ctx.streaming_metrics.clone(),
+                "Join",
+            );
+            pin_mut!(aligned_stream);
+            #[for_await]
+            for message in aligned_stream {
+                match message? {
+                    AlignedMessage::Left(chunk) => {
+                        lhs_buffer = Some(chunk);
+                    }
+                    AlignedMessage::Right(chunk) => {
+                        rhs_buffer = Some(chunk);
+                    }
+                    AlignedMessage::Barrier(barrier) => {
+                        if lhs_buffer.is_none() && rhs_buffer.is_none() {
+                            yield Message::Barrier(barrier);
+                            continue;
+                        }
+                        let Some(lhs_chunk) = lhs_buffer.take() else {
+                            bail!("lhs buffer should not be empty ");
+                        };
+                        let Some(rhs_chunk) = rhs_buffer.take() else {
+                            bail!("rhs buffer should not be empty ");
+                        };
+
+                        if !(1..=2).contains(&lhs_chunk.cardinality()) {
+                            bail!("lhs chunk cardinality should be 1 or 2");
+                        }
+                        if !(1..=2).contains(&rhs_chunk.cardinality()) {
+                            bail!("rhs chunk cardinality should be 1 or 2");
+                        }
+                        if lhs_chunk.cardinality() != rhs_chunk.cardinality() {
+                            bail!("lhs and rhs chunk cardinality should be the same");
+                        }
```

Comment:
Is this assumption a bit too strong? What we know at this operator is that logically, the relations it inputs have only one row each, but we do not assume that only one chunk of changes will come during a barrier period, or that this chunk contains only one or two rows. Can we still parse each row of every incoming chunk when we receive it to get the current values on both sides? This way, we can handle chunks like (`-+-+-+`).

Reply:
The semantics of row merge is such that the input chunks should only have 1 operation, e.g. update, insert or delete.

Update will be for chunks with cardinality 2.
Insert will be for chunks with cardinality 1.

It's specific to simple normal agg and simple approx percentile agg, where the output cardinality of both sides should be 1 (insert or delete) or 2 (update) at most.

We can generalize it later if needed.

Reply:
Oh I got what you mean. We can't simply overwrite the chunks.

Reply:
Handled by https://github.com/risingwavelabs/risingwave/pull/17930/commits/677edab74870da69ee9ff4374e6c28f5da688de9